services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -36,7 +36,26 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
     }
   }
   override def getFiles(project: Project, path: Path)(implicit ec: ExecutionContext): AsyncResult[List[String]] = ???
-  override def getProjectVersions(project: Project)(implicit ec: ExecutionContext): AsyncResult[Project] = ???
+
+  override def getProjectVersions(project: Project)(implicit ec: ExecutionContext): AsyncResult[Seq[Version]] =
+    if (project.repository == null)
+      Future.failed(VersioningException("There is no repository in this project"))
+    else {
+      val versionsListUrl: String = s"${config.url}projects/${project.repository}/repository/tags"
+      httpClient
+        .get(url = versionsListUrl, headers = config.token)
+        .map(
+          resp =>
+            if (resp.status != StatusCodes.OK.intValue)
+              Left(VersioningException(s"Could not take versions. Response status: ${resp.status}"))
+            else {
+              val versionsBody = Json.parse(resp.body).validate[Seq[Version]]
+              Right(versionsBody.get)
```

Comment:
Almost forgot.
You must check whether `versionsBody` contains `JsSuccess` or `JsError` using pattern matching instead of simple `.get`. If it contains `JsError` then you must return `Left` instead of `Right`.

Reply:
Fixed. But I ask you to check, I use "try" and "as" structures. Is it ok?

Reply:
Good enough, but in my opinion, you can use `.validate` + `.fold` or `.validate` + pattern matching to handle this situation without throwing an exception. Take a look at the implementation of the `.as` method. You can do a similar thing.

Reply:
Fixed to validate.